primitives/src/v1/mod.rs

```diff
@@ -1349,9 +1349,41 @@ pub struct DisputeStatementSet {
 	pub statements: Vec<(DisputeStatement, ValidatorIndex, ValidatorSignature)>,
 }
 
+impl AsRef<DisputeStatementSet> for DisputeStatementSet {
+	fn as_ref(&self) -> &DisputeStatementSet {
+		&self
+	}
+}
+
 /// A set of dispute statements.
 pub type MultiDisputeStatementSet = Vec<DisputeStatementSet>;
 
+/// A _checked_ set of dispute statements.
+#[derive(Clone)]
+pub struct CheckedDisputeStatementSet(DisputeStatementSet);
+
+impl AsRef<DisputeStatementSet> for CheckedDisputeStatementSet {
+	fn as_ref(&self) -> &DisputeStatementSet {
+		&self.0
+	}
+}
+
+impl Into<DisputeStatementSet> for CheckedDisputeStatementSet {
+	fn into(self) -> DisputeStatementSet {
+		self.0
+	}
+}
+
+impl CheckedDisputeStatementSet {
+	/// Convert from an unchecked, checking _must_ be done in the context.
+	pub fn from_unchecked(unchecked: DisputeStatementSet) -> Self {
```

Comment:
Why doesn't impl `From` work here ?

Reply:
Would it make sense to perform the checks here and return an `Option`?

Reply:
Technically yes, but that would require additional context and enforce another change in behavior which I wanted to avoid as much as possible. The changeset is already too large.

Reply:
I would at least call it `unsafe_from_unchecked` or something. Being able to grep through source code for `unsafe` to manually check things the compiler won't check is a nice property to have.

Reply:
Not really keen on using `unsafe` here, this conveys a different meaning in rust-context. Up for alt-names though :bike: :house:

Reply:
Quite the opposite - I find keyword re-use more dangerous, it conflates memory safety and topological safety under a single name/tag `unsafe`. This was discussed in various groups (rust security SIG iirc besides some design lib PRs) and was agreed to not be a good idea. This is just the same thing, so we should not go down this path imho.

Reply:
We eventually want to follow design of `Signed<_>` where the conversion requires checking and only retain the current `fn from_unchecked` for tests.